src/fparser/statements.py

```diff
@@ -1748,47 +1750,169 @@ class Case(Statement):
-        return s
-    def analyze(self): return
+            txt += ' ' + self.name
+        return txt
+
+    def analyze(self):
+        return
+
+
+class TypeIs(Statement):
+    """
+    TYPE IS <type-selector> [ <case-constract-name> ]
+    <type-selector> = ( <type-value-range-list> )
+    <type-value-range> = <case-value>
+    <case-value> = <char>
+    """
+    match = re.compile(r'type\b\s*is\b\s*\(.*\)\s*\w*\Z', re.I).match
+
+    def process_item(self):
+        ''' Populate the state of this object by parsing the associated
+        line of code '''
+        line = self.item.get_line()[4:].lstrip()
+        idx = line.find('(')
+        if idx > 0:
+            line = line[idx:]
+            idx = line.find(')')
+            items = split_comma(line[1:idx].strip(), self.item)
+            line = line[idx+1:].lstrip()
+        for idx in range(len(items)):
+            itm = self.item.copy(items[idx])
+            rlst = []
+            for rpart in itm.get_line().split(':'):
+                rlst.append(itm.apply_map(rpart.strip()))
+            items[idx] = rlst
+        self.items = items
+        self.name = line
+        parent_name = getattr(self.parent, 'name', '')
+        if self.name and self.name != parent_name:
+            self.warning(
+                'expected type-is-construct-name %r but got %r, skipping.'
+                % (parent_name, self.name))
+            self.isvalid = False
+        return
+
+    def tofortran(self, isfix=None):
+        ''' Create the Fortran representation of this object and return
+        it as a string '''
+        tab = self.get_indent_tab(isfix=isfix)
+        text = tab + 'TYPE IS'
+        if self.items:
+            lst = []
+            for item in self.items:
+                lst.append((' : '.join(item)).strip())
+            text += ' ( %s )' % (', '.join(lst))
+        else:
+            raise ParseError("TYPE IS construct must have arguments")
+        if self.name:
+            text += ' ' + self.name
+        return text
+
+    def analyze(self):
+        return
+
+
+class ClassIs(Statement):
+    """
+    CLASS <class-selector>
+    <class-selector> = ( IS <type-value-range-list> | DEFAULT )
+    <type-value-range> = <case-value>
+    <case-value> = <char>
+    """
+    match = re.compile(r'class\b\s*(is\b\s*\(.*\)|default)\s*\w*\Z',
+                       re.I).match
+
+    def process_item(self):
+        ''' Populate the state of this object by parsing the string '''
+        line = self.item.get_line()[5:].lstrip()
+        if line.lower().startswith('is'):
+            # We have a 'class is ...' statement
+            idx = line.find('(')
+            line = line[idx:]
+            idx = line.find(')')
+            items = split_comma(line[1:idx].strip(), self.item)
+            line = line[idx+1:].lstrip()
+        else:
+            # We have a 'class default' statement
+            assert line.lower().startswith('default'), repr(line)
```

Comment:
*Sighs heavily* :-) assert's are used liberally through the code but I agree, it should really raise an exception (although any such exception is caught at a higher level and logged). Since I've touched this code I'll change it...